gradio/oauth.py

```diff
@@ -105,18 +105,39 @@ async def oauth_redirect_callback(request: fastapi.Request) -> RedirectResponse:
             # repeatedly. Since cookies cannot get bigger than 4kb, the token will be truncated at some point - hence
             # losing the state. A workaround is to delete the cookie and redirect the user to the login page again.
             # See https://github.com/lepture/authlib/issues/622 for more details.
-            login_uri = "/login/huggingface"
-            if "_target_url" in request.query_params:
-                login_uri += (
-                    "?"
-                    + urllib.parse.urlencode(  # Keep same _target_url as before
-                        {"_target_url": request.query_params["_target_url"]}
-                    )
-                )
+
+            # Delete all keys that are related to the OAuth state, just in case
             for key in list(request.session.keys()):
-                # Delete all keys that are related to the OAuth state
                 if key.startswith("_state_huggingface"):
                     request.session.pop(key)
+
+            # Parse query params
+            nb_redirects = int(request.query_params.get("_nb_redirects", 0))
+            target_url = request.query_params["_target_url"]
+
+            # Build /login URI with the same query params as before and bump nb_redirects count
+            login_uri = "/login/huggingface"
+            login_uri += "?" + urllib.parse.urlencode(
+                {"_nb_redirects": nb_redirects + 1}
+            )
+            if "_target_url" in request.query_params:
+                # Keep same _target_url as before
+                login_uri += "&" + urllib.parse.urlencode({"_target_url": target_url})
+
+            # If the user is redirected more than 3 times, it is very likely that the cookie is not working properly.
+            # (e.g. browser is blocking third-party cookies in iframe). In this case, redirect the user in the
+            # non-iframe view.
+            if nb_redirects > 2:
```

Comment:
nit: consider defining a constant MAX_REDIRECTS
```suggestion
            if nb_redirects > MAX_REDIRECTS:
```

Reply:
addressed in https://github.com/gradio-app/gradio/pull/8951/commits/51b0bac3c33fa9988d81126f2dc1d6a4fe08523b